AxonDeepSeg/train_network.py

```diff
@@ -54,588 +58,178 @@ def train_model(path_trainingset, path_model, config, path_model_init=None,
-    :param prediction: Tensor, the prediction outputted by the network. Shape (N,H,W,C).
-    :param gt: Tensor, the gold standard we work with. Shape (N,H,W,C).
-    :return: Vector, dice per class for the current batch.
-    """
+    # Adam Optimizer for Unet
+    adam = keras.optimizers.Adam(lr=learning_rate, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0)
+
+    # Compile the model with Categorical Cross Entropy loss and Adam Optimizer
+    model.compile(optimizer=adam, loss="categorical_crossentropy",
+                  metrics=["accuracy", dice_axon, dice_myelin, dice_coef])
+
+    train_steps = len(train_ids) // batch_size
+    valid_steps = len(valid_ids) // batch_size
+
+    ########################## Use Checkpoints to save best Acuuracy and Loss ###########
+
+    # Save the checkpoint in the /models/path_model folder
+    filepath_acc = str(path_model) + "/ADS-best_acc.hdf5"
 
-    sum_ = tf.reduce_sum(prediction, axis=1) + tf.reduce_sum(gt, axis=1)
-    intersection = tf.logical_and(tf.cast(prediction, tf.bool), tf.cast(gt, tf.bool))
-    return tf.multiply(2., tf.div(tf.reduce_sum(tf.cast(intersection, tf.float32), axis=1), sum_))
+    # Keep only a single checkpoint, the best over test accuracy.
+    checkpoint_acc = ModelCheckpoint(filepath_acc,
+                                     monitor='val_acc',
+                                     verbose=0,
+                                     save_best_only=True,
+                                     mode='max')
+
+    # Save the checkpoint in the /models/path_model folder
+    filepath_loss = str(path_model) + "/ADS-best_loss.hdf5"
+    print(filepath_loss)
+
+    # Keep only a single checkpoint, the best over test loss.
+    checkpoint_loss = ModelCheckpoint(filepath_loss,
+                                      monitor='val_loss',
+                                      verbose=0,
+                                      save_best_only=True,
+                                      mode='min')
+
+
+
+    ########################## Use Checkpoints to save best Acuuracy and Loss ###########
+    model.fit_generator(train_generator, validation_data=(valid_generator), steps_per_epoch=train_steps,
+                        validation_steps=valid_steps,
+                        epochs=epochs, callbacks=[tensorboard, checkpoint_loss, checkpoint_acc])
+
+    ########################## Save the model after Training ###########
+
+    model.save(str(path_model) + '/model.hdf5')
+
+
+    # Add ops to save and restore all the variables.
+    saver = tf.train.Saver()
+
+    # Save Model in ckpt format
+    custom_objects = {'dice_axon': dice_axon, 'dice_myelin': dice_myelin, 'dice_coef': dice_coef}
+    model = load_model(str(path_model) + "/model.hdf5", custom_objects=custom_objects)
+
+    sess = K.get_session()
+    # Save the model to be used by TF framework
+    save_path = saver.save(sess, str(path_model) + "/model.ckpt")
+    print(save_path)
+
+
+# Defining the Loss and  Performance Metrics
```

Comment:
The for functions below should have docstrings explaining what's being calculated. Also, there seems to be a lot of repetition, i.e. dice myelin and dic axon are the same except for a single value (1 or 2). There should just be one dice function, and  either a flag for myelin/axon or have those information be parsed out before calling the function.

Reply:
A set of unit tests for these functions would also be warranted to ensure the actual behaviour matches the expected behaviour.